### DISCUSSION TITLE
Understanding ETIMEDOUT logs

### DISCUSSION THREAD

**Discussion by rarkins** [sindresorhus/got#2125] [Q&A]
If we see a 60000ms timeout logged like either of the below:

```
    "code": "ETIMEDOUT",
    "timings": {
      "start": 1661933000729,
      "socket": 1661933000945,
      "lookup": 1661933108895,
      "error": 1661933108935,
      "phases": {
        "wait": 216,
        "dns": 107950,
        "total": 108206
      }
    },
```

or

```
    "code": "ETIMEDOUT",
    "timings": {
      "start": 1661931980128,
      "socket": 1661931980332,
      "error": 1661932081516,
      "phases": {
        "wait": 204,
        "total": 101388
      }
    },
```

Does this imply that the problem is occurring very early, e.g. during DNS lookup?

And considering that the host is `api.github.com` and there are other concurrent requests succeeding at the same time.. can we conclude anything from this?

Also is there any `got` timeout settings we could adjust to fail this faster and retry?

**Comment by viceice** [comment]
We're still stuck got v11[^1]

[^1]: https://github.com/renovatebot/renovate/blob/7c056534e7e85c4b230a2a5177d55e680fc62b7b/package.json#L180

**Reply by viceice** [reply]
maybe related to
- https://github.com/sindresorhus/got/discussions/1605

**Comment by rarkins** [comment]
@williamcoates you seemed to have similar problems last year. Were you able to resolve them?

**Comment by szmarczak** [comment]
https://github.com/szmarczak/http-timer

Looking at the first one, it seems that it took `107.95` seconds to get the reply from DNS. The latter timed out after 101 seconds with no DNS reply.

I suggest that you set the [`lookup`](https://github.com/sindresorhus/got/blob/main/documentation/6-timeout.md#lookup) timeout to `500` or `1000`.

> there are other concurrent requests succeeding at the same time.

You could try using https://github.com/szmarczak/cacheable-lookup

**Reply by viceice** [reply]
currently trying this 🙃 
- https://github.com/renovatebot/renovate/pull/17533

**Reply by rarkins** [reply]
Is there any way to confirm that the request was retried? According to the docs an ETIMEDOUT would be retried twice by default? Here's what I see:

```
  "err": {
    "name": "TimeoutError",
    "code": "ETIMEDOUT",
    "timings": {
      "start": 1662100033130,
      "socket": 1662100033130,
      "lookup": 1662100139603,
      "connect": 1662100139603,
      "error": 1662100139646,
      "phases": {
        "wait": 0,
        "dns": 106473,
        "tcp": 0,
        "total": 106516
      }
    },
    "event": "request",
    "message": "Timeout awaiting 'request' for 60000ms",
    "stack": "RequestError: Timeout awaiting 'request' for 60000ms\n    at ClientRequest.<anonymous> (/home/ubuntu/renovateapp/node_modules/got/dist/source/core/index.js:970:65)\n    at Object.onceWrapper (events.js:520:26)\n    at ClientRequest.emit (events.js:412:35)\n    at ClientRequest.emit (domain.js:475:12)\n    at ClientRequest.origin.emit (/home/ubuntu/renovateapp/node_modules/@szmarczak/http-timer/dist/source/index.js:43:20)\n    at TLSSocket.socketErrorListener (_http_client.js:475:9)\n    at TLSSocket.emit (events.js:400:28)\n    at TLSSocket.emit (domain.js:475:12)\n    at emitErrorNT (internal/streams/destroy.js:106:8)\n    at emitErrorCloseNT (internal/streams/destroy.js:74:3)\n    at processTicksAndRejections (internal/process/task_queues.js:82:21)\n    at runNextTicks (internal/process/task_queues.js:64:3)\n    at processTimers (internal/timers.js:497:9)\n    at Timeout.timeoutHandler [as _onTimeout] (/home/ubuntu/renovateapp/node_modules/got/dist/source/core/utils/timed-out.js:36:25)\n    at listOnTimeout (internal/timers.js:559:11)\n    at processTimers (internal/timers.js:500:7)",
    "options": {
      "headers": {
        "user-agent": "Renovate Bot (GitHub App 2740)",
        "accept": "application/vnd.github.machine-man-preview+json",
        "authorization": "***********",
        "content-type": "application/json",
        "content-length": "623",
        "accept-encoding": "gzip, deflate, br"
      },
      "url": "https://api.github.com/graphql",
      "hostType": "github-releases",
      "username": "",
      "password": "",
      "method": "POST",
      "http2": false
    }
  }
```

I also find it confusing why the lookup seems to take an even 100s nearly every time. If it tried once that would be 60s, or retried once 120s, etc?

**Reply by szmarczak** [reply]
You can use [`response.retryCount`](https://github.com/sindresorhus/got/blob/main/documentation/3-streams.md#retrycount-1) or `error.response.retryCount`.

**Reply by szmarczak** [reply]
> I also find it confusing why the lookup seems to take an even 100s nearly every time. If it tried once that would be 60s, or retried once 120s, etc?

The timings are per single request (aka: a single try), they do not accumulate.

**Reply by szmarczak** [reply]
Is this with the cache turned on?

**Reply by szmarczak** [reply]
If so, you could try using `8.8.8.8` or `1.1.1.1` like this: https://github.com/szmarczak/cacheable-lookup/blob/5b08a3610a5ec1328da583a6327111428ff85c42/tests/test.js#L465-L481

**Reply by rarkins** [reply]
We just added a custom cache written by @viceice: https://github.com/renovatebot/renovate/blob/main/lib/util/http/dns.ts

We haven't tried a custom DNS server yet. Sorry but I don't understand how to based on the test case you linked to.

Is there a way to set the lookup timeout to be low while keeping everything else to 60s? I want to try lowering them one by one to avoid unforeseen problems.

**Reply by szmarczak** [reply]
```diff
export const dnsCache = new CacheableLookup({
  cache: new QuickLRU({ maxSize: 1000 }),
});

+dnsCache.servers = ['1.1.1.1', '1.0.0.1', '8.8.8.8', '8.8.4.4'];
```

**Reply by szmarczak** [reply]
> Is there a way to set the lookup timeout to be low while keeping everything else to 60s? I want to try lowering them one by one to avoid unforeseen problems.

```js
await got({
	...,
	timeout: {
		request: 60 * 1000,
		lookup: 1000,
	}
});
```

**Reply by rarkins** [reply]
Because we're not using `CacheableLookup` directly, I'm not sure that code snippet works for us. IIRC the reason for a custom cache was because GitHub returned a zero second TTL, which meant it wasn't cached at all.

**Reply by szmarczak** [reply]
What's the GitHub domain?

**Reply by rarkins** [reply]
api.github.com

**Reply by szmarczak** [reply]
Are you sure `api.github.com` has 0 TTL? Here it returns 60.

**Comment by rarkins** [comment]
We switched to http2 so maybe now skip the lookup stage, but still fail:

```
  "err": {
    "name": "TimeoutError",
    "code": "ETIMEDOUT",
    "timings": {
      "start": 1662177845433,
      "socket": 1662177845567,
      "upload": 1662177946760,
      "error": 1662177946789,
      "phases": {
        "wait": 134,
        "request": null,
        "total": 101356
      }
    },
    "event": "request",
    "message": "Timeout awaiting 'request' for 60000ms",
```

What does the above indicate if the URL is a POST to https://api.github.com/graphql. Does "upload" mean the POST submission essentially times out and the response hasn't even started?

**Reply by szmarczak** [reply]
https://github.com/szmarczak/http-timer#timerrequest

It means that it finished uploading, but the server didn't reply.

**Reply by rarkins** [reply]
Doesn't it indicate that upload finished 100s after socket? That sounds more like an incomplete upload than a completed one though, or do I misunderstand it?

**Reply by szmarczak** [reply]
Yeah, it looks like it finished writing 100s after the request had started. Do note that it's Node.js streams. It just means that the last chunk was flushed 100s after the request had started. E.g. if you send a chunk of 1GB it will be flushed, but the data won't be all sent at once. I guess you don't do that, since it says 100 seconds and that took some time.

> That sounds more like an incomplete upload than a completed one though

Why is that?

**Reply by rarkins** [reply]
I find it hard to believe that the upload stage would take almost exactly 100s each time this happens. It's just a graphql POST with small data, so seems unlikely that it coincidentally takes 100s time after time. More likely there's an error in timings reporting or something else I'm missing

**Reply by rarkins** [reply]
Before we enabled http2, it was the lookup phase timing out or reporting 100s. Seems like something else is going wrong

**Reply by szmarczak** [reply]
> I find it hard to believe that the upload stage would take almost exactly 100s each time this happens.

Hm... you're right, something's definitely wrong. Are you able to reproduce the issue locally?

**Reply by rarkins** [reply]
Unfortunately we can't reproduce the problem locally. It's only occurring in our hosted app, which is Node.js 14/Ubuntu 20.04 running inside Docker containers on EC2 on AWS. Happens to certain repos consistently but not always. Typically ones needing a lot of GitHub graphql API queries. Similar problem with some large (10s of MB) npm registry fetches. Both scenarios it seems like querying multiple in parallel is to blame. Not sure if there's memory pressure, socket exhaustion, or simply a low level bug where they overwrite each other. 

Any thoughts on what we could debug log at times when we catch such timeouts?

**Reply by szmarczak** [reply]
Definitely please try Node.js 16 & 18. HTTP/2 in Node.js <16 is extremely buggy. The timings could be also incorrect due to Node.js stream bugs (e.g. it could say that it finished uploading when it actually didn't), not sure if this is the case though. Let me know what's the result.

**Reply by rarkins** [reply]
FYI the timeouts have been happening since before we enabled http2, so it seems unrelated.

**Reply by viceice** [reply]
@rarkins should we try node v16 anyways?

**Reply by szmarczak** [reply]
> FYI the timeouts have been happening since before we enabled http2, so it seems unrelated.

I'm aware of that, however it could possibly reveal the underlying issue. I think there could be a deadlock somewhere, but without trying we can't do anything.

**Comment by rarkins** [comment]
Here's a summary of what we found so far:
- We fixed the GitHub API GraphQL timeouts by rolling back to an earlier solution which uses REST and no caching tricks. We don't know yet if the problem was GraphQL-related or our tricks, but will eventually confirm that as we add it back step by step
- Enabling http2 with Node 14 for npmjs resulted in strange problems, including ECONNRESET. Reverting http2 fixed them. We have subsequently upgraded to Node 16 but not yet tested http2 again

**Reply by szmarczak** [reply]
> We fixed the GitHub API GraphQL timeouts by rolling back to an earlier solution which uses REST and no caching tricks. We don't know yet if the problem was GraphQL-related or our tricks, but will eventually confirm that as we add it back step by step

Can you send me a commit link? I'm curious what has changed.

> Enabling http2 with Node 14 for npmjs resulted in strange problems, including ECONNRESET. Reverting http2 fixed them. We have subsequently upgraded to Node 16 but not yet tested http2 again

Please do try it out and let me know.

**Reply by rarkins** [reply]
This was the PR where we changed our github tags/releases fetching: https://github.com/renovatebot/renovate/pull/17685
